### ISSUE TITLE
Hello author, thank you for your contribution. I encountered some issues while using this code.

### ISSUE THREAD

**Issue by zky0701** [shaoyanpan/Synthetic-CT-generation-from-MRI-using-3D-transformer-based-denoising-diffusion-model#13]
Hello author, thank you for your contribution. I encountered some issues while using this code. During the first round of validation, I found that the validation speed was very slow, taking almost five hours per epoch. What is the reason for this? The data I am using is in NII format, and this is my code. We look forward to your reply!

def evaluate(model, epoch, path, data_loader1, best_loss):
    model.eval()
    prediction = []
    true = []
    img = []
    loss_all = []
    aa = time.time()
   
    with torch.no_grad():
        # Wrap the data_loader1 with tqdm for evaluation
        for i, (x1, y1) in enumerate(tqdm(data_loader1, total=len(data_loader1), desc="Evaluation Progress", ncols=100)):
            target = y1.to(device)
            condition = x1.to(device)
            
            with torch.cuda.amp.autocast():
                sampled_images = inferer(condition, diffusion_sampling, model)
            
            loss = metric(sampled_images, target)
            print(f'L1 loss: {loss.item()}')
            
            # img.append(x1.cpu().numpy())
            # true.append(target.cpu().numpy())
            # prediction.append(sampled_images.cpu().numpy())
            # loss_all.append(loss.cpu().numpy())
            img.append(x1.to(device))  # Keep the image on GPU
            true.append(target.to(device))  # Keep the target on GPU
            prediction.append(sampled_images.to(device))  # Keep the prediction on GPU
            loss_all.append(loss.to(device))  # Keep the loss on GPU

        
        print(f'Optimization time: {(time.time() - aa):.2f}s')
        
        # Save the results to .nii.gz using NiBabel
        # Assuming img, true, prediction are numpy arrays that are 3D or 4D
        for idx, (image, pred, truth) in enumerate(zip(img, prediction, true)):
            # Assuming the shape is (C, H, W) or (C, H, W, D)
            # You may need to adjust the shape depending on your data dimensions
            
            # Save input image (condition)
            img_nii = nib.Nifti1Image(image[0], np.eye(4))  # Assuming 3D image and identity matrix as affine
            nib.save(img_nii, path + f'input_image_{epoch}_{idx}.nii.gz')
            
            # Save predicted image
            pred_nii = nib.Nifti1Image(pred[0], np.eye(4))  # Assuming 3D image and identity matrix as affine
            nib.save(pred_nii, path + f'predicted_image_{epoch}_{idx}.nii.gz')
            
            # Save ground truth
            truth_nii = nib.Nifti1Image(truth[0], np.eye(4))  # Assuming 3D image and identity matrix as affine
            nib.save(truth_nii, path + f'ground_truth_{epoch}_{idx}.nii.gz')

        # If the current loss is the best, save the best result
        if np.mean(loss_all) < best_loss:
            print('Saving the best model')
            # Save the best prediction results as .nii.gz
            best_data = {"prediction": prediction, 'loss': np.mean(loss_all)}
            # You might want to process it into a single volume or multiple volumes, depending on your setup
            best_pred_nii = nib.Nifti1Image(np.array(prediction), np.eye(4))
            nib.save(best_pred_nii, path + f'best_prediction_epoch{epoch}.nii.gz')
        
        return np.mean(loss_all)
![123](https://github.com/user-attachments/assets/e773ba23-022f-43e9-97e3-fd7176455c53)

**Comment by shaoyanpan** [comment]
Hi, I think it is hard to see since I don't have the code. Based opn your information, 46 training samples only takes 158 seconds. Now if the whole img_size = (256,256,128) and patch_size = (64,64,4), you need to process 2x2x32 patches for each image. Now, Idealy, one sample in validation (if you use 50 timesteps for generatio), you need 158/46*50x(2x2x32), about 20000 seconds.

The first thing I suggest you is to modify:
overlap = 0.5 (under the "building the testing function")
Lower the overlap will significantly improve the speed. Consider the above calcuation is for overlap=0.

Second:
num_channels=64 (under the "Build the MC-IDDPM network")
Lower the num_channels reduce the size of network, improve speed

Third:
patch_size = (64,64,4)
Increase the patch_size (e.g., 128,128,32) will really improve your testing speed.

**Comment by zky0701** [comment]
> Hi, I think it is hard to see since I don't have the code. Based opn your information, 46 training samples only takes 158 seconds. Now if the whole img_size = (256,256,128) and patch_size = (64,64,4), you need to process 2x2x32 patches for each image. Now, Idealy, one sample in validation (if you use 50 timesteps for generatio), you need 158/46*50x(2x2x32), about 20000 seconds.
> 
> The first thing I suggest you is to modify: overlap = 0.5 (under the "building the testing function") Lower the overlap will significantly improve the speed. Consider the above calcuation is for overlap=0.
> 
> Second: num_channels=64 (under the "Build the MC-IDDPM network") Lower the num_channels reduce the size of network, improve speed
> 
> Third: patch_size = (64,64,4) Increase the patch_size (e.g., 128,128,32) will really improve your testing speed.

Thank you very much for taking the time to answer my question and for providing me with suggestions for improvement! I have another question. How long does it take for each evaluation during your training? I need about an hour and a half, is this time normal? Looking forward to your answer!

**Comment by shaoyanpan** [comment]
Actually I remember my dataset has MRI size of 192x256x32, and patch size of 128x128x4. So I need 4 (time for each patch)x1.5x2x8 = about 96 seconds for each scan, using A6000 GPU. But I think, if you don't want the optimized accuracy, you can use 5 generation timestep so to reduce the time to 9.6 seconds in my case. Trust me, timestep of 5 would give you, myabe a little bit worse, but not significant worse result (you probably cannot even see the difference).

**Comment by zky0701** [comment]
> Actually I remember my dataset has MRI size of 192x256x32, and patch size of 128x128x4. So I need 4 (time for each patch)x1.5x2x8 = about 96 seconds for each scan, using A6000 GPU. But I think, if you don't want the optimized accuracy, you can use 5 generation timestep so to reduce the time to 9.6 seconds in my case. Trust me, timestep of 5 would give you, myabe a little bit worse, but not significant worse result (you probably cannot even see the difference).

Okay, I understand now, and thank you for taking the time to answer my question!
